grudge/models/__init__.py

```diff
@@ -49,13 +49,22 @@ def max_characteristic_velocity(self, t, fields, dcoll):
 
     def estimate_rk4_timestep(self, dcoll, t=None, fields=None):
         """Estimate the largest stable timestep for an RK4 method."""
-        from grudge.dt_utils import (dt_non_geometric_factor,
+        from grudge.dt_utils import (dt_non_geometric_factors,
                                      dt_geometric_factors)
+        from meshmode.dof_array import DOFArray
         import grudge.op as op
 
+        actx = dcoll._setup_context
         max_lambda = self.max_characteristic_velocity(t, fields, dcoll)
-        dt_factor = \
-            (dt_non_geometric_factor(dcoll)
-             * op.nodal_min(dcoll, "vol", dt_geometric_factors(dcoll)))
+        # Scale each group array of geometric factors by the corresponding
+        # group non-geometric factor
+        dt_factors = DOFArray(
+            actx,
+            data=(
+                cng * geo_facts
+                for cng, geo_facts in zip(dt_non_geometric_factors(dcoll),
+                                          dt_geometric_factors(dcoll))
+            )
+        )
 
-        return dt_factor * (1 / max_lambda)
+        return op.nodal_min(dcoll, "vol", dt_factors * (1 / max_lambda))
```

Comment:
I'd honestly feel better if this were a function rather than a method. Otherwise, it starts forcing objects into places where they kind of have no business being. If that means killing the `HyperbolicOperator` class, that's probably for the better. (mirgecom uses functions for its operators, and I think that's by and large a good thing.)

Reply:
I haven't completely removed this function, but is this closer to what you're thinking: https://github.com/inducer/grudge/pull/121/commits/a3659999b0e5dcfa57de21c1090e4a706b675629?

Reply:
Yep, that looks excellent!